Enable a long translog retention policy by default

bleskes opened the pull request:
#25147  added the translog deletion policy but didn't enable it by default. This PR enables a default retention of 512MB (same maximum size of the current translog) and an age of 12 hours (i.e., after 12 hours all translog files will be deleted). This increases to chance to have an ops based recovery, even if the primary flushed or the replica was offline for a few hours.

In order to see which parts of the translog are committed into lucene the translog stats are extended to include information about uncommitted operations.

Views now include all translog ops and guarantee, as before, that those will not go away. Snapshotting a view allows to filter out generations that are not relevant based on a specific sequence number.

I still have to write some docs and add a migration note, but I think we can start reviewing.

Relates to #10708

s1monw commented on core/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -606,6 +608,9 @@ public synchronized void updateMetaData(final IndexMetaData metadata) {
             if (refreshTask.getInterval().equals(indexSettings.getRefreshInterval()) == false) {
                 rescheduleRefreshTasks();
             }
+            if (trimTranslogTask.getInterval().equals(indexSettings.getTranslogRetentionCheckInterval()) == false) {
```
Can you elaborate why you go down this route instead of using `IndexShard#afterWriteOperation()` and maybe also do it in `IndexShard#flush`? I'd love to rather remove tasks than adding them?! it's just way easier to reason about things if they are happening due to a user interaction form the outside

bleskes replied:
> Can you elaborate why you go down this route instead of using IndexShard#afterWriteOperation() 

This is not instead but on top of it. The problem is that if people have a 12h retention policy and  we need to start cleaning stuff 12hs later (assuming they index full speed and roll over to a new index).

>  I'd love to rather remove tasks than adding them?! 

I'm totally with you. I spent some time thinking about ways to avoid the regular check but everything I came up with ended up more complex. I considered stuff scheduling things in in activity etc.. always come back to this being the clearest/easiest. I'll be more than happy to hear about alternatives.

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -885,13 +885,25 @@ public CompletionStats completionStats(String... fields) {
-        engine.getTranslog().rollGeneration();
+        final Translog translog = engine.getTranslog();
+        translog.rollGeneration();
+        translog.trimUnreferencedReaders();
```
hmm you run `translog.trimUnreferencedReaders();` under the engine lock [here](https://github.com/elastic/elasticsearch/pull/25294/files#diff-74a25d57bbb52202a04fe43d8632172eR1321)  but you run it without a lock here so I wonder when we have to run under lock and if we can assert on that?

bleskes replied:
That's a great observation. Strictly speaking we don't need to coordinate through the write lock here. In the place you linked to I was mostly interested in the error handling and that we fail the engine if anything goes wrong (as we close the translog on error as well). I was worried for poisonous situations where we keep trying to get stuff from a closed translog.  I will work on removing the read lock in the engine and fold this functionality (trim and roll) into the engine so we can be consistent. agreed?

s1monw replied:
I think the readlock is healthy here, I wonder if we would rather fold this part into the engine and don't call it from the outside and add a parameter `boolean rollGeneration` to the trim method?

jasontedor replied:
I too find it weird that we have `IndexShard#trimTranslog` that goes through the engine (with error handling) but ultimately amounts to trimming unreferenced readers, and we have this method that does the same without going through the engine instead poking all the way through to the translog.

s1monw commented on core/src/main/java/org/elasticsearch/index/translog/Translog.java:
```diff
@@ -560,14 +594,33 @@ public Snapshot newSnapshot(long minGeneration) {
     }
 
+    private Stream<? extends BaseTranslogReader> readersAboveMinSeqNo(long minSeqNo) {
+        assert readLock.isHeldByCurrentThread() || writeLock.isHeldByCurrentThread();
```
please have one assert per check it's easier to see what failed

bleskes replied:
I'm confused - this is an or check. I will add a message.

s1monw replied:
nevermind message is good

s1monw commented on core/src/main/java/org/elasticsearch/index/translog/Translog.java:
```diff
@@ -1522,6 +1576,10 @@ public void trimUnreferencedReaders() throws IOException {
                 // we're shutdown potentially on some tragic event, don't delete anything
                 return;
             }
+            if (readers.isEmpty()) {
```
maybe instead of adding this intermediate return we can invert and only do the work if it's not empty?

bleskes replied:
yeah, I can do that and thought about it. The reason I went with this is that to do something like `shouldTrimTranslog` where I can add this shortcut, I will also have to go and ask the retention policy for the current trimming situation and add tests etc. Since this method is not called on every write, but I rather on every roll / flush/ background check every 10m.. I felt this was simpler. Let me know what you think.

s1monw replied:
methods should have a single return statement IMO, everything else is more complex in 99% of the cases.

s1monw commented on core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java:
```diff
@@ -154,4 +153,8 @@ private long getMinTranslogGenRequiredByViews() {
     }
+
+    synchronized long getViewCount(long viewGen) {
+        return translogRefCounts.getOrDefault(viewGen, Counter.newCounter(false)).get();
```
you are creating a new counter anyway even if it's there I think we can safe that object and use `computeIfAbsent`?

bleskes replied:
++. will change

s1monw commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -128,13 +129,14 @@ public RecoverySourceHandler(final IndexShard shard, RecoveryTargetHandler recov
 
             if (isSequenceNumberBasedRecoveryPossible) {
                 logger.trace("performing sequence numbers based recovery. starting at [{}]", request.startingSeqNo());
+                startingSeqNo = request.startingSeqNo();
```
the changes in here are unrelated to the trimming code, I wonder if we can do it in 2 steps and add the integration in a followup, that way reviews are much simpler and more effective?

bleskes replied:
yes, this became less important than what it used to be after some iterations. That said the change is mostly related to reporting and that acquiring a snapshot now has a seq# in it's constructor (to avoid scanning a large translog). I can try to roll that part back but I was afraid it will change the system too much (we will stream more ops from the translog). I'm not sure it's worth it tbh.

s1monw replied:
`+727 −336` I think we should try harder to make them more contained. really!

bleskes commented:
@s1monw thanks. I responded to some of you comment. Will apply code changes tomorrow.

jasontedor commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -1317,6 +1317,24 @@ public CommitId flush(boolean force, boolean waitIfOngoing) throws EngineExcepti
 
+    @Override
+    public void trimTranslog() throws EngineException {
+        try (ReleasableLock lock = readLock.acquire()) {
```
Would you mind renaming this to `ignored`? It keeps the IDE from complaining, and it makes it clear that the resource is intentionally unused through the scope of the try-with-resources block.

jasontedor commented on core/src/main/java/org/elasticsearch/index/translog/Translog.java:
```diff
@@ -560,14 +594,33 @@ public Snapshot newSnapshot(long minGeneration) {
+            .filter(reader -> {
+                final long maxSeqNo = reader.getCheckpoint().maxSeqNo;
+                return maxSeqNo == SequenceNumbersService.UNASSIGNED_SEQ_NO ||
+                    maxSeqNo >= minSeqNo;
```
Nit: I think this fits on the preceding line without wrapping.

jasontedor commented on core/src/main/java/org/elasticsearch/index/translog/TranslogStats.java:
```diff
@@ -30,20 +31,29 @@
     private long translogSizeInBytes;
     private int numberOfOperations;
+    private long uncommittedSizeInBytes;
+    private int  uncommittedOperations;
```
I wonder if since these are summed in node stats if this should be a long (to reduce the change of overflow)?

bleskes replied:
I agree, but I think we need to do this consistently and I didn't want to touch the other `numberOfOperations` field (and the bwc code that entails) in the same PR. That's why I did it like this.

jasontedor replied:
To be clear, you're saying that you'll address this in a follow-up?

bleskes replied:
~can~ will do.

bleskes commented:
@s1monw I took out all the scheduled trimming (we can decide what to do about it later) and did the engine changes. 

@jasontedor I addressed your feedback. 

Can you please look again?

bleskes commented:
Thx @jasontedor @s1monw for the iterations.
